File: infra/pulumi/pkg/model/defaults.go
package model

import (
	"encoding/binary"
	"net"
	"strconv"
	"strings"
)

func ApplyProviderDefaults(spec *ClusterSpec) *ClusterSpec {
	switch CanonicalProviderName(spec.Provider) {
	case "", "aws":
		spec = applyAwsDefaults(spec)
	case "openstack":
		spec = applyOpenStackDefaults(spec)
	case "gcp":
		spec = applyGcpDefaults(spec)
	case "azure":
		spec = applyAzureDefaults(spec)
	case "alibaba":
		spec = applyAlibabaDefaults(spec)
	case "proxmox":
		spec = applyProxmoxDefaults(spec)
	case "oci":
		spec = applyOciDefaults(spec)
	case "digitalocean":
		spec = applyDigitalOceanDefaults(spec)
	default:
		return spec
	}
	spec.Provider = CanonicalProviderName(spec.Provider)
	spec.SubnetCidrs = ensureDefaultSubnets(spec.VpcCidr, spec.SubnetCidrs, requiredSubnetCount(spec.Provider))
	// MasterCount cross-cutting default — 모든 provider 공통. 0 (unset) 이면 single master.
	// 짝수 (2, 4) 는 etcd quorum 의 split-brain 위험이라 odd 로 강제 (3, 5).
	if spec.MasterCount <= 0 {
		spec.MasterCount = 1
	} else if spec.MasterCount%2 == 0 {
		spec.MasterCount = spec.MasterCount + 1
	}
	// root 디스크 cross-cutting default — 모든 provider 공통 50GB (k8s 노드 최소 권장).
	// 0/음수(unset)면 50 으로 정규화. 기본 ~8GB AMI/이미지로는 컨테이너 런타임 + 로그만으로
	// kubelet ephemeral-storage 임계 초과 → NodeHasDiskPressure / eviction.
	if spec.RootDiskSizeGb <= 0 {
		spec.RootDiskSizeGb = 50
	}
	return spec
}

func CanonicalProviderName(provider string) string {
	switch strings.ToLower(strings.TrimSpace(provider)) {
	case "", "aws":
		return "aws"
	case "gcp", "google", "googlecloud":
		return "gcp"
	case "azure", "msazure":
		return "azure"
	case "alibaba", "alicloud", "aliyun":
		return "alibaba"
	case "openstack", "open-stack":
		return "openstack"
	case "proxmox", "proxmoxve", "pve":
		return "proxmox"
	case "oci", "oracle", "oraclecloud", "oraclecloudinfrastructure":
		return "oci"
	case "digitalocean", "digital-ocean", "do":
		return "digitalocean"
	default:
		return strings.ToLower(strings.TrimSpace(provider))
	}
}

func ResolvedOsImage(spec *ClusterSpec) string {
	switch CanonicalProviderName(spec.Provider) {
	case "openstack":
		return spec.OpenstackImageName
	case "gcp":
		return "ubuntu-2404-lts"
	case "azure":
		return "Canonical Ubuntu 24.04 LTS"
	case "alibaba":
		return "Ubuntu 24.04"
	case "proxmox", "oci", "digitalocean":
		return "Ubuntu 24.04"
	default:
		return "ubuntu-24.04"
	}
}

func applyAwsDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "aws"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-demo"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.42.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "t3.large"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "t3.large"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	if spec.Database.Name == "" {
		spec.Database.Name = "anycloud"
	}
	if spec.Database.Username == "" {
		spec.Database.Username = "anycloud"
	}
	if spec.Database.InstanceClass == "" {
		spec.Database.InstanceClass = "db.t4g.micro"
	}
	if spec.Database.AllocatedStorageGb == 0 {
		spec.Database.AllocatedStorageGb = 20
	}
	return spec
}

func applyOpenStackDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "openstack"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-openstack"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.90.0.0/24"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	if spec.OpenstackImageName == "" {
		spec.OpenstackImageName = "ubuntu-24.04"
	}
	if spec.OpenstackFlavorName == "" {
		spec.OpenstackFlavorName = "m1.large"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = spec.OpenstackFlavorName
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = spec.OpenstackFlavorName
	}
	return spec
}

func applyGcpDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "gcp"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-gcp"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.52.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "e2-standard-2"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "e2-standard-2"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	return spec
}

func applyAzureDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "azure"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-azure"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.62.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "Standard_D4s_v5"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "Standard_D4s_v5"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	if spec.AzureResourceGroup == "" {
		spec.AzureResourceGroup = spec.Name + "-rg"
	}
	return spec
}

func applyAlibabaDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "alibaba"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-alibaba"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.72.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "ecs.g6.large"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "ecs.g6.large"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	return spec
}

func applyProxmoxDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "proxmox"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-proxmox"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.84.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "proxmox-standard-2x4"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "proxmox-standard-2x4"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	return spec
}

func applyOciDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "oci"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-oci"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.86.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "ubuntu"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "VM.Standard.E4.Flex"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "VM.Standard.E4.Flex"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	return spec
}

func applyDigitalOceanDefaults(spec *ClusterSpec) *ClusterSpec {
	if spec.Provider == "" {
		spec.Provider = "digitalocean"
	}
	if spec.Name == "" {
		spec.Name = "anycloud-digitalocean"
	}
	if spec.Environment == "" {
		spec.Environment = "dev"
	}
	if spec.VpcCidr == "" {
		spec.VpcCidr = "10.88.0.0/16"
	}
	if spec.SSHUser == "" {
		spec.SSHUser = "root"
	}
	if spec.MasterInstanceType == "" {
		spec.MasterInstanceType = "s-2vcpu-4gb"
	}
	if spec.WorkerInstanceType == "" {
		spec.WorkerInstanceType = "s-2vcpu-4gb"
	}
	if spec.WorkerCount == 0 {
		spec.WorkerCount = 2
	}
	if spec.KubernetesVersion == "" {
		spec.KubernetesVersion = "1.31"
	}
	if spec.PodCidr == "" {
		spec.PodCidr = "192.168.0.0/16"
	}
	if spec.ServiceCidr == "" {
		spec.ServiceCidr = "10.96.0.0/12"
	}
	if tok, err := EnsureJoinToken(spec.JoinToken); err == nil {
		spec.JoinToken = tok
	}
	return spec
}

func requiredSubnetCount(provider string) int {
	switch CanonicalProviderName(provider) {
	case "aws":
		return 2
	default:
		return 1
	}
}

func ensureDefaultSubnets(vpcCidr string, existing []string, required int) []string {
	if required <= 0 {
		required = 1
	}
	subnets := append([]string{}, existing...)
	if len(subnets) >= required {
		return subnets
	}

	generated := generateSubnets(vpcCidr, required)
	if len(subnets) == 0 {
		return generated
	}

	for _, cidr := range generated {
		if len(subnets) >= required {
			break
		}
		if !containsString(subnets, cidr) {
			subnets = append(subnets, cidr)
		}
	}
	return subnets
}

func generateSubnets(vpcCidr string, count int) []string {
	ip, network, err := net.ParseCIDR(vpcCidr)
	if err != nil || ip == nil || network == nil {
		return nil
	}

	baseIP := ip.To4()
	maskSize, bits := network.Mask.Size()
	if baseIP == nil || bits != 32 {
		return nil
	}

	subnetMaskSize := maskSize
	if subnetMaskSize < 24 {
		subnetMaskSize = 24
	}
	if subnetMaskSize > 30 {
		subnetMaskSize = maskSize
	}

	step := uint32(1) << uint32(32-subnetMaskSize)
	base := binary.BigEndian.Uint32(baseIP.Mask(network.Mask))
	subnets := make([]string, 0, count)
	for i := 0; i < count; i++ {
		cidrIP := make(net.IP, 4)
		binary.BigEndian.PutUint32(cidrIP, base+(uint32(i)*step))
		subnets = append(subnets, cidrIP.String()+"/"+itoa(subnetMaskSize))
	}
	return subnets
}

func containsString(items []string, target string) bool {
	for _, item := range items {
		if item == target {
			return true
		}
	}
	return false
}

func itoa(value int) string {
	return strconv.Itoa(value)
}

File: libs/cluster-provisioning-spring-boot-starter/src/main/java/io/aipaas/cluster/provisioning/program/Defaults.java
package io.aipaas.cluster.provisioning.program;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Provider 별 default 적용 + cross-cutting 보정.
 *
 * <p>caller 가 제공한 ClusterSpec 의 빈 필드를 provider 의 권장 값으로 채워 새 record 반환.
 * cross-cutting default (masterCount odd 강제 — etcd quorum split-brain 방지, rootDiskSizeGb≥50 — k8s
 * NodeHasDiskPressure 방지) 도 본 메서드가 적용.
 */
public final class Defaults {

    private Defaults() {}

    private static final String DEFAULT_ENVIRONMENT = "dev";
    private static final String DEFAULT_K8S_VERSION = "1.31";
    private static final String DEFAULT_POD_CIDR = "192.168.0.0/16";
    private static final String DEFAULT_SERVICE_CIDR = "10.96.0.0/12";
    private static final int DEFAULT_WORKER_COUNT = 2;
    private static final int DEFAULT_ROOT_DISK_GB = 50;

    /**
     * Provider 별 default 값 테이블. 새 CSP 추가 시 entry 1줄 + (선택) {@link #applyProviderSpecific}
     * 의 case 추가.
     */
    private record ProviderDefaults(
            String name,
            String vpcCidr,
            String masterInstanceType,
            String workerInstanceType,
            String sshUser) {}

    private static final Map<String, ProviderDefaults> TABLE = Map.of(
            "aws",          new ProviderDefaults("anycloud-demo",         "10.42.0.0/16", "t3.large",            "t3.large",            "ubuntu"),
            "gcp",          new ProviderDefaults("anycloud-gcp",          "10.52.0.0/16", "e2-standard-2",       "e2-standard-2",       "ubuntu"),
            "azure",        new ProviderDefaults("anycloud-azure",        "10.62.0.0/16", "Standard_D4s_v5",     "Standard_D4s_v5",     "ubuntu"),
            "alibaba",      new ProviderDefaults("anycloud-alibaba",      "10.72.0.0/16", "ecs.g6.large",        "ecs.g6.large",        "ubuntu"),
            "openstack",    new ProviderDefaults("anycloud-openstack",    "10.90.0.0/24", null,                  null,                  "ubuntu"),
            "oci",          new ProviderDefaults("anycloud-oci",          "10.86.0.0/16", "VM.Standard.E4.Flex", "VM.Standard.E4.Flex", "ubuntu"),
            "digitalocean", new ProviderDefaults("anycloud-digitalocean", "10.88.0.0/16", "s-2vcpu-4gb",         "s-2vcpu-4gb",         "root"));

    public static ClusterSpec applyProviderDefaults(ClusterSpec raw) {
        String canonical = ProviderName.canonical(raw.provider());
        ProviderDefaults pd = TABLE.get(canonical);
        if (pd == null) return raw.toBuilder().provider(canonical).build();

        ClusterSpec.Builder b = raw.toBuilder()
                .provider(canonical)
                .name(blankOr(raw.name(), pd.name()))
                .environment(blankOr(raw.environment(), DEFAULT_ENVIRONMENT))
                .vpcCidr(blankOr(raw.vpcCidr(), pd.vpcCidr()))
                .sshUser(blankOr(raw.sshUser(), pd.sshUser()))
                .workerCount(raw.workerCount() == 0 ? DEFAULT_WORKER_COUNT : raw.workerCount())
                .kubernetesVersion(blankOr(raw.kubernetesVersion(), DEFAULT_K8S_VERSION))
                .podCidr(blankOr(raw.podCidr(), DEFAULT_POD_CIDR))
                .serviceCidr(blankOr(raw.serviceCidr(), DEFAULT_SERVICE_CIDR))
                .joinToken(JoinTokens.ensure(raw.joinToken()));

        if (pd.masterInstanceType() != null) {
            b.masterInstanceType(blankOr(raw.masterInstanceType(), pd.masterInstanceType()))
                    .workerInstanceType(blankOr(raw.workerInstanceType(), pd.workerInstanceType()));
        }
        applyProviderSpecific(canonical, raw, b);

        ClusterSpec withProvider = b.build();
        List<String> subnets = ensureDefaultSubnets(
                withProvider.vpcCidr(), withProvider.subnetCidrs(), requiredSubnetCount(canonical));

        int masterCount = withProvider.masterCount();
        if (masterCount <= 0) masterCount = 1;
        else if (masterCount % 2 == 0) masterCount = masterCount + 1;

        int rootDisk = withProvider.rootDiskSizeGb() <= 0 ? DEFAULT_ROOT_DISK_GB : withProvider.rootDiskSizeGb();

        return withProvider.toBuilder()
                .subnetCidrs(subnets)
                .masterCount(masterCount)
                .rootDiskSizeGb(rootDisk)
                .build();
    }

    /** CSP-specific 필드 (Azure RG / OpenStack image+flavor / AWS database) — table 으로 표현 어려운 case. */
    private static void applyProviderSpecific(String canonical, ClusterSpec raw, ClusterSpec.Builder b) {
        switch (canonical) {
            case "aws" -> b.database(applyDbDefaults(raw.database(), "anycloud", "anycloud", "db.t4g.micro", 20));
            case "azure" -> {
                String name = blankOr(raw.name(), TABLE.get("azure").name());
                b.azureResourceGroup(blankOr(raw.azureResourceGroup(), name + "-rg"));
            }
            case "openstack" -> {
                String flavor = blankOr(raw.openstackFlavorName(), "m1.large");
                b.masterInstanceType(blankOr(raw.masterInstanceType(), flavor))
                        .workerInstanceType(blankOr(raw.workerInstanceType(), flavor))
                        .openstackImageName(blankOr(raw.openstackImageName(), "ubuntu-24.04"))
                        .openstackFlavorName(flavor);
            }
            default -> { /* no extras */ }
        }
    }

    public static String resolvedOsImage(ClusterSpec spec) {
        return switch (ProviderName.canonical(spec.provider())) {
            case "openstack" -> spec.openstackImageName();
            case "gcp" -> "ubuntu-2404-lts";
            case "azure" -> "Canonical Ubuntu 24.04 LTS";
            case "alibaba", "oci", "digitalocean" -> "Ubuntu 24.04";
            default -> "ubuntu-24.04";
        };
    }

    private static DatabaseSpec applyDbDefaults(
            DatabaseSpec db, String name, String username, String instanceClass, int storageGb) {
        if (db == null) return DatabaseSpec.disabled();
        if (!db.enabled()) return db;
        return new DatabaseSpec(
                true,
                blankOr(db.name(), name),
                blankOr(db.username(), username),
                db.password(),
                blankOr(db.instanceClass(), instanceClass),
                db.allocatedStorageGb() == 0 ? storageGb : db.allocatedStorageGb(),
                db.publiclyAccessible());
    }

    private static String blankOr(String value, String fallback) {
        return (value == null || value.isBlank()) ? fallback : value;
    }

    private static int requiredSubnetCount(String provider) {
        return "aws".equals(provider) ? 2 : 1;
    }

    /** IPv4 VPC CIDR 안의 subnet block 들을 균등 분할로 생성. existing 가 부족하면 generated 로 보강. */
    static List<String> ensureDefaultSubnets(String vpcCidr, List<String> existing, int required) {
        List<String> subnets = new ArrayList<>(existing == null ? List.of() : existing);
        if (subnets.size() >= required) return subnets;
        List<String> generated = generateSubnets(vpcCidr, required);
        if (subnets.isEmpty()) return generated;
        for (String cidr : generated) {
            if (subnets.size() >= required) break;
            if (!subnets.contains(cidr)) subnets.add(cidr);
        }
        return subnets;
    }

    static List<String> generateSubnets(String vpcCidr, int count) {
        if (vpcCidr == null) return List.of();
        int slash = vpcCidr.indexOf('/');
        if (slash < 0) return List.of();
        String ipStr = vpcCidr.substring(0, slash);
        int maskSize;
        try {
            maskSize = Integer.parseInt(vpcCidr.substring(slash + 1));
        } catch (NumberFormatException e) {
            return List.of();
        }
        long ip = parseIpv4(ipStr);
        if (ip < 0) return List.of();

        int subnetMaskSize = Math.max(maskSize, 24);
        if (subnetMaskSize > 30) subnetMaskSize = maskSize;
        long step = 1L << (32 - subnetMaskSize);
        long mask = maskSize == 0 ? 0L : (0xFFFFFFFFL << (32 - maskSize)) & 0xFFFFFFFFL;
        long base = ip & mask;

        List<String> subnets = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            long current = base + (long) i * step;
            subnets.add(formatIpv4(current) + "/" + subnetMaskSize);
        }
        return subnets;
    }

    private static long parseIpv4(String s) {
        String[] octets = s.split("\\.");
        if (octets.length != 4) return -1;
        long v = 0;
        for (String octet : octets) {
            int o;
            try {
                o = Integer.parseInt(octet);
            } catch (NumberFormatException e) {
                return -1;
            }
            if (o < 0 || o > 255) return -1;
            v = (v << 8) | o;
        }
        return v;
    }

    private static String formatIpv4(long v) {
        return ((v >> 24) & 0xFF) + "." + ((v >> 16) & 0xFF) + "." + ((v >> 8) & 0xFF) + "." + (v & 0xFF);
    }
}
